File: c/args.c
#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "args.h"

void print_help() {
    printf(
"an application for running \"train\" simulations of a givne acceleration\n"
"profile that will calculate the final velocity and position of the train\n"
"\n"
"Usage: sim [OPTIONS] <PATH>\n"
"\n"
"Arguments:\n"
"  <PATH> the input file to load acceleration data from\n"
"\n"
"Options:\n"
"  -t, --threads <THREADS> specifies the number of threads to use for\n"
"                          simulation\n"
"  -a, --algo <ALGO>       specifies the summation algorithm to use for the\n"
"                          simulation [default: left-riemann]\n"
"  -i, --iterations <ITER> specifies the number times to run the program, for\n"
"                          benchmarking purposes\n"
"  -s, --step <STEP>       specifies the number of steps to take in bewteen each\n"
"                          summation calculation [default: 10]\n"
    );

}

void print_full_help() {
    printf(
"an application for running \"train\" simulations of a givne acceleration\n"
"profile that will calculate the final velocity and position of the train\n"
"\n"
"Usage: sim [OPTIONS] <PATH>\n"
"\n"
"Arguments:\n"
"  <PATH> the input file to load acceleration data from\n"
"\n"
"Options:\n"
"\n"
"  -t, --threads <THREADS>\n"
"        specifies the number of threads to use for simulation\n"
"\n"
"  -a, --algo <ALGO>\n"
"        specifies the summation algorithm to use for the simulation\n"
"        [default: left-riemann] [possible-values: left-riemann, mid-riemann,\n"
"        right-riemann, trapezoidal, simpsons]\n"
"\n"
"  -i, --iterations <ITER>\n"
"        specifies the number times to run the program, for benchmarking purposes\n"
"\n"
"  -s, --step <STEP>\n"
"        specifies the number of steps to take in bewteen each summation\n"
"        calculation [default: 10]\n"
    );
}

int32_t app_args_init(struct app_args* self, int argc, char** argv) {
    self->file_path = NULL;
    self->sim.threads = 1;
    self->sim.algo = 0;
    self->sim.step = 10;
    self->sim.iterations = 1;

    struct option long_options[] = {
        {"threads", required_argument, 0, 0 },
        {"step", required_argument, 0, 0 },
        {"iterations", required_argument, 0, 0 },
        {"algo", required_argument, 0, 0 },
        {"help", no_argument, 0, 0}
    };

    int32_t option_index = 0;

    while (1) {
        int32_t c = getopt_long(argc, argv, "t:s:i:a:h:", long_options, &option_index);

        if (c == -1) {
            return 0;
        }

        switch (c) {
        case 0:
            switch (option_index) {
            case 0:
                if (parse_threads_arg(optarg, &self->sim.threads) != 0) {
                    return 1;
                }
                break;
            case 1:
                if (parse_step_arg(optarg, &self->sim.step) != 0) {
                    return 1;
                }
                break;
            case 2:
                if (parse_iterations_arg(optarg, &self->sim.iterations) != 0) {
                    return 1;
                }
                break;
            case 3:
                if (parse_algo_arg(optarg, &self->sim.algo) != 0) {
                    return 1;
                }
                break;
            case 4:
                print_full_help();

                return 1;
            }
            break;
        case 't':
            if (parse_threads_arg(optarg, &self->sim.threads) != 0) {
                return 1;
            }
            break;
        case 's':
            if (parse_step_arg(optarg, &self->sim.step) != 0) {
                return 1;
            }
            break;
        case 'i':
            if (parse_iterations_arg(optarg, &self->sim.iterations) != 0) {
                return 1;
            }
            break;
        case 'a':
            if (parse_algo_arg(optarg, &self->sim.algo) != 0) {
                return 1;
            }
            break;
        case 'h':
            print_help();

            return 1;
        case '?':
            break;
        default:
            break;
        }
    }

    return 0;
}

int32_t parse_threads_arg(const char* arg, int32_t* threads) {
    long parsed = 0;

    if (parse_l(arg, &parsed) != 0) {
        fprintf(stderr, "invalid thread size provided\n");

        return 1;
    }

    if (parsed > INT32_MAX || parsed < 1) {
        fprintf(stderr, "invalid thread size provided\n");

        return 1;
    }

    *threads = parsed;

    return 0;
}

int32_t parse_algo_arg(const char* arg, int32_t* algo) {
    if (strncmp(arg, "left-riemann", 13) == 0) {
        *algo = LEFT_RIEMANN;
    } else if (strncmp(arg, "mid-riemann", 12) == 0) {
        *algo = MID_RIEMANN;
    } else if (strncmp(arg, "right-riemann", 14) == 0) {
        *algo = RIGHT_RIEMANN;
    } else if (strncmp(arg, "trapezoidal", 12) == 0) {
        *algo = TRAPEZOIDAL;
    } else if (strncmp(arg, "simpsons", 9) == 0) {
        *algo = SIMPSONS;
    } else {
        fprintf(stderr, "invalid algo provided\n");

        return 1;
    }

    return 0;
}

int32_t parse_step_arg(const char* arg, int32_t* step) {
    long parsed = 0;

    if (parse_l(arg, &parsed) != 0) {
        fprintf(stderr, "invalid step provided\n");

        return 1;
    }

    if (parsed > INT32_MAX || parsed < 1) {
        fprintf(stderr, "invalid step provided\n");

        return 1;
    }

    *step = parsed;

    return 0;
}

int32_t parse_iterations_arg(const char* arg, int32_t* iterations) {
    long parsed = 0;

    if (parse_l(arg, &parsed) != 0) {
        fprintf(stderr, "invalid iterations provided\n");

        return 1;
    }

    if (parsed > INT32_MAX || parsed < 1) {
        fprintf(stderr, "invalid iterations provided\n");

        return 1;
    }

    *iterations = parsed;

    return 0;

}

int parse_l(const char* str, long* value) {
    char* endptr;
    *value = strtol(str, &endptr, 10);

    if (*endptr != '\0') {
        return -1;
    }

    if ((*value == LONG_MAX && errno == ERANGE) || errno == EINVAL) {
        return -1;
    }

    return 0;
}

File: src/args.rs
use std::path::PathBuf;
use std::str::FromStr;

use anyhow::Context;
use clap::{Args, Parser, Subcommand, ValueEnum};

use crate::summation::{self, InterpolateLookup};

/// an application for running "train" simulations of a given acceleration
/// profile that will calculate the final velocity and position of the train
#[derive(Debug, Parser)]
pub struct App {
    /// specifies the number of threads to use for calculations
    #[arg(short, long, default_value("1"))]
    pub threads: usize,

    #[command(flatten)]
    pub opts: SimOpts,

    #[command(subcommand)]
    pub sim: SimKind,
}

/// common options between simulations
#[derive(Debug, Clone, Args)]
pub struct SimOpts {
    /// specifies the summation algorithm to use for the simulation
    #[arg(short, long, default_value("left-riemann"))]
    pub algo: AppAlgo,

    /// determines the number of times to run the program, for benchmarking
    /// purposes
    #[arg(short, long, default_value("100"))]
    pub iterations: u32,

    /// specifies the amount of steps to take in between each summation
    /// calculation
    #[arg(short, long, default_value("100"))]
    pub step: u32,
}

/// the available summation algorithms that the simulation is capable of
/// running
#[derive(Debug, Clone, ValueEnum)]
pub enum AppAlgo {
    LeftRiemann,
    MidRiemann,
    RightRiemann,
    Trapezoidal,
    Simpsons,
}

/// the different kins of simulations available for the program to run
///
/// currently the only supported kind is loading data from a csv file
#[derive(Debug, Subcommand)]
pub enum SimKind {
    /// runs a simulation from a given acceleration profile
    Csv(CsvSim),
}

/// options for running a simulation from a specified csv file
#[derive(Debug, Args)]
pub struct CsvSim {
    /// loads acceleration data in a specific column from the csv file
    #[arg(long)]
    pub column: Option<String>,

    /// the csv file path to load
    pub path: PathBuf,
}

impl CsvSim {
    /// retrieves the path of the specified csv file
    ///
    /// if the given path is relative then it will be resolved using the
    /// current working directory
    fn get_path(&self) -> anyhow::Result<PathBuf> {
        if self.path.is_relative() {
            let cwd =
                std::env::current_dir().context("failed to retrieve current working directory")?;

            Ok(cwd.join(&self.path))
        } else {
            Ok(self.path.clone())
        }
    }

    /// builds the [`csv::Reader`] from the provided csv path
    fn get_csv_reader(&self) -> anyhow::Result<csv::Reader<std::fs::File>> {
        let path = self.get_path()?;

        let mut builder = csv::ReaderBuilder::new();

        if self.column.is_some() {
            builder.has_headers(true);
        } else {
            builder.has_headers(false);
        }

        builder.from_path(&path).context("failed to load csv file")
    }

    /// parses the given csv file into a lookup table that supports
    /// interpolation
    pub fn get_callable(self) -> anyhow::Result<summation::InterpolateLookup> {
        let mut rtn = Vec::new();
        let mut reader = self.get_csv_reader()?;

        let data_index = if let Some(column) = self.column {
            let mut maybe_index: Option<usize> = None;
            let headers = reader.headers().context("failed to retrieve csv headers")?;

            for (index, header) in headers.iter().enumerate() {
                if header == column {
                    maybe_index = Some(index);

                    break;
                }
            }

            maybe_index.context("failed to find the desired csv column")?
        } else {
            0
        };

        let records = reader.records();

        for (index, try_record) in records.enumerate() {
            let record = try_record
                .with_context(|| format!("failed to retrieve csv entry. {}", index + 1))?;

            let value = record
                .get(data_index)
                .with_context(|| format!("failed to retrieve csv entry column. {}", index + 1))?;

            rtn.push(f64::from_str(value).with_context(|| {
                format!("failed to convert csv entry into float. {}", index + 1)
            })?);
        }

        Ok(InterpolateLookup::from(rtn))
    }
}
